README
Java & C# implemenation of K-D tree by
Prof. Simon D. Levy
Associate Professor and Department Head
Computer Science Department
407 Parmly Hall Washington & Lee University
Lexington, VA 24450
-----------------------------------------------

Installing this JAR into local mvn repo--> mvn install:install-file -DgroupId=omegha.kdtree -DartifactId=kdtree -Dversion=1.0 -Dpackaging=jar -DgeneratePom=true -Dfile=kd.jar

A KD-tree is a data structure for efficient search and nearest-neighbor(s) computation of points in K-dimensional space. I like programming in Java and couldn't find any Java KD-tree implementations on the Web, so I wrote this one. (There appear to be better implementations now, like the ones mentioned on this site.) I also thank the following contributors:

    Eric Burnett, who patched a memory leak and suggested an improvement for memory allocation.

    Bjoern Heckel, who added the ability to return multiple nearest neighbors, instead of just one.

    Andrey Fomin, who has provided a Polish translation of this page.

    Michael Lorton, who added unit tests, changed the interface to use generics and modern container classes, and made the code thread-safe (though as with all aspects of software, we make no guarantees). If you want to use the unit tests, you will have to obtain the junit package from Sourceforge. Michael is now using KDTree to help match people on his SpeedDate website. 

The file kd.jar contains the Java class edu.wlu.cs.levy.CG.KDTree, which supports the standard KD-tree operations insert, delete, equality search, range search, and nearest-neighbor. There is also a version for C#, written by Marco A Alvarez, in KDTreeDLL.zip.

The JAR file contains the class files of course, but also all the source and the Makefile, for anyone wishing to hack the code (KDTree.nearest could probably be improved). The demonstration programs kddemo, kdtime, kdrange, and kdnbrs show you how to use the KDTree class. To run these programs, do the following in Unix (or the equivalent Java commands on your platform);

% javac -classpath .:kd.jar kddemo.java
% javac -classpath .:kd.jar kdrange.java
% javac -classpath .:kd.jar kdtime.java
% javac -classpath .:kd.jar kdnbrs.java
% java -classpath .:kd.jar kddemo
% java -classpath .:kd.jar kdrange
% java -classpath .:kd.jar kdtime 10000 2 100 
                                 (for example)                               
% java -classpath .:kd.jar kdnbrs 10000 3 100 
                                 (for example)                               

Demonstration programs can be found under examples/ directory
References ->http://home.wlu.edu/~levys/software/kd/

Please send comments, questions, and suggestions to levys@wlu.edu.   Like all software I've posted, KDTree is released under the Gnu Lesser General Public License. 



CHANGELOG.md
# Changelog

+ removed superfluous sqrt
+ added size()
+ support for Checker object
+ made exception public
+ fixed range-search to exclude deleted values
+ added tests
+ made delete optional
+ changed to use generic lists
+ removed unnecessary "SDL" step
+ made thread-safe
+ supported editing of found objects
+ adding timeout


src/test/java/edu/wlu/cs/levy/CG/KDTreeTest.java
/*
 * Copyright (C) 2015 Simon Levy
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package edu.wlu.cs.levy.CG;

import java.util.List;
import org.junit.Assert;
import org.junit.Test;

/**
 *
 * @author MSL
 */
public class KDTreeTest {

    static java.util.Random rand = new java.util.Random();

    static double[] makeSample(int dims) {
        double[] rv = new double[dims];
        for (int j = 0; j < dims; ++j) {
            rv[j] = rand.nextDouble();
        }
        return rv;
    }

    static double distSquared(double[] p0, double[] p1) {
        double rv = 0;
        for (int i = 0; i < p0.length; i++) {
            double diff = p0[i] - p1[i];
            rv += (diff * diff);
        }
        return rv;
    }

    @Test
    public void testNearestNeighborList() {
        NearestNeighborList<String> nnl = new NearestNeighborList<String>(3);
        nnl.insert("A", 3.0);
        nnl.insert("B", 2.0);
        nnl.insert("D", 0.0);
        nnl.insert("C", 1.0);

        Assert.assertEquals(2.0, nnl.getMaxPriority(), 0.1);
        Assert.assertEquals("B", nnl.getHighest());
        Assert.assertEquals("B", nnl.removeHighest());
        Assert.assertEquals("C", nnl.removeHighest());
        Assert.assertEquals("D", nnl.removeHighest());
    }

    @Test
    public void testNearestNeighbor() throws KDException {
        int dims = 3;
        int samples = 300;
        KDTree<Integer> kt = new KDTree<Integer>(dims);
        double[] targ = makeSample(dims);

        int min_index = 0;
        double min_value = Double.MAX_VALUE;
        for (int i = 0; i < samples; ++i) {
            double[] keys = makeSample(dims);
            kt.insert(keys, new Integer(i));

            /*
             for the purposes of test, we want the nearest EVEN-NUMBERED point
             */
            if ((i % 2) == 0) {
                double dist = distSquared(targ, keys);
                if (dist < min_value) {
                    min_value = dist;
                    min_index = i;
                }
            }
        }

        List<Integer> nbrs = kt.nearest(targ, 1, new Checker<Integer>() {
            public boolean usable(Integer v) {
                return (v.intValue() % 2) == 0;
            }
        });

        Assert.assertEquals(1, nbrs.size());
        if (nbrs.size() == 1) {
            Assert.assertEquals(min_index, nbrs.get(0).intValue());
        }
    }

    @Test
    public void testRange() throws KDException {
        int dims = 2;
        KDTree<Object> kt = new KDTree<Object>(dims);
        double[] p0 = {0.5, 0.5};
        double[] p1 = {0.65, 0.5};
        double[] p2 = {0.75, 0.5};

        kt.insert(p0, new Object());
        kt.insert(p1, new Object());
        kt.insert(p2, new Object());

        double[] lower = {0.25, 0.3};
        double[] upper = {0.7, 0.6};

        List<Object> rv = kt.range(lower, upper);
        Assert.assertEquals(2, rv.size());

        kt.delete(p1);
        rv = kt.range(lower, upper);
        Assert.assertEquals(1, rv.size());
    }

    @Test
    public void testSearch() throws KDException {
        int dims = 3;
        int samples = 300;
        KDTree<Object> kt = new KDTree<Object>(dims);
        double[] targ = makeSample(dims);
        Object treasure = new Object();
        kt.insert(targ, treasure);

        for (int i = 0; i < samples; ++i) {
            double[] keys = makeSample(dims);
            kt.insert(keys, new Integer(i));
        }

        Object found = kt.search(targ);
        Assert.assertSame(treasure, found);

        kt.delete(targ);
        found = kt.search(targ);
        Assert.assertNull(found);

    }

    @Test
    public void testDelete() throws KDException {
        int dims = 3;
        KDTree<Object> kt = new KDTree<Object>(dims);
        double[] targ = makeSample(dims);
        kt.insert(targ, new Object());
        kt.delete(targ);
        try {
            kt.delete(targ);
            Assert.assertFalse(true);
        } catch (edu.wlu.cs.levy.CG.KeyMissingException e) {
            // supposed to be here
        }
        kt.delete(targ, true);
        Assert.assertEquals(0, kt.size());
    }

    @Test
    public void testEditing() throws KDException {
        int dims = 3;
        KDTree<Object> kt = new KDTree<Object>(dims);
        double[] targ = makeSample(dims);

        Object p1 = "p1";
        Object p2 = "p2";
        kt.insert(targ, p1);
        try {
            kt.insert(targ, p2);
            Assert.assertFalse(true);
        } catch (edu.wlu.cs.levy.CG.KeyDuplicateException e) {
            // supposed to be here
        }

        kt.edit(targ, new Editor.OptionalInserter<Object>(p2));
        Object found = kt.search(targ);
        Assert.assertSame(p1, found);
        kt.edit(targ, new Editor.Replacer<Object>(p2));
        found = kt.search(targ);
        Assert.assertSame(p2, found);
    }

}


examples/java/kddemo.java
// illustrates behavior of KDTree class

import edu.wlu.cs.levy.CG.KDTree;

class kddemo {

    public static void main(String [] args) {

	double [] A = {2, 5};
	double [] B = {1, 1};
	double [] C = {3, 9};
	double [] T = {1, 10};
	
	// make a KD-tree and add some nodes
	KDTree<String> kd = new KDTree<String>(2);
	try {
	    kd.insert(A, new String("A"));
	    kd.insert(B, new String("B"));
	    kd.insert(C, new String("C"));
	}
	catch (Exception e) {
	    System.err.println(e);
	}


	// look for node B
	try {
	    String n = kd.search(B);
	    System.err.println(n);

	}
	catch (Exception e) {
	    System.err.println(e);
	}

	try {

	    // find T's nearest neighbor, which should be C
	    String n = kd.nearest(T);
	    System.err.println(n);

	    // remove C from the tree
	    kd.delete(C);

	    // now T's nearest neighbor should be A
	    n = kd.nearest(T);
	    System.err.println(n);
	}
	catch (Exception e) {
	    System.err.println(e);
	}
    }

}


examples/java/kddist.java
// Illustrates distance-based nearest-neighbors method of KDTree class.  Creates 
// a KDTree of M keys, and finds neighbors of D-dimensional point in center of 
// space with distance Hamming distance H (all coords = 0.5), with H, D, M 
// specified on command line.

import edu.wlu.cs.levy.CG.KDTree;

import java.util.List;
import java.util.Iterator;

class kddist {

    public static void main(String [] args) {

	java.util.Random r = new java.util.Random(0);

	if (args.length < 3) {
	    System.err.print("Usage: java kddist <# points> <# dims>");
	    System.err.println(" <distance>");
	    System.exit(1);
	}
	
	int m = Integer.parseInt(args[0]);
	int d = Integer.parseInt(args[1]);
	double v = Double.parseDouble(args[2]);

	double [][] keys = new double [m][d];

	double [] targ = new double [d];
	for (int k=0; k<d; ++k) {
	    targ[k] = 0.5;
	}

	// make a D-dimensional KD-tree
	KDTree<Integer> kd = new KDTree<Integer>(d);

	try {

	    // add M randomly keyed nodes

	    for (int i=0; i<m; ++i) {
		for (int j=0; j<d; ++j) {
		    keys[i][j] = r.nextDouble();
		}
		kd.insert(keys[i], i);
	    }

	    // get nearest neighbors and show their keys

	    long start = time();

	    List<Integer> nbrs = kd.nearestHamming(targ, v);
	
	    System.err.println((time() - start) + " msec");

	    for (int j : nbrs) {
		for (int k=0; k<d; ++k) {
		    System.out.print(keys[j][k] + " ");
		}
		System.err.println("\t" + dist(keys[j], targ));
	    }
	}
	catch (Exception e) {
	    System.err.println(e);
	}
    }	
    
    private static double dist(double [] a, double [] b) {
	double d = 0;
	for (int k=0; k<a.length; ++k) {
	    double s = a[k] - b[k];
	    d += Math.abs(s);
	}
	return d;
    }

    private static long time() {
	java.util.GregorianCalendar cal = new java.util.GregorianCalendar();
	return cal.getTimeInMillis();
    }



}


examples/java/kdnbrs.java
// Illustrates n-nearest-neighbors method of KDTree class.  Creates a KDTree
// of M keys, and finds N neighbors of D-dimensional point in center of space
// (all coords = 0.5), with N, D, M specified on command line.

import edu.wlu.cs.levy.CG.KDTree;

import java.util.List;
import java.util.Iterator;

class kdnbrs {

    public static void main(String [] args) {

	java.util.Random r = new java.util.Random(0);

	if (args.length < 3) {
	    System.err.print("Usage: java kdnbrs <# points> <# dims>");
	    System.err.println(" <# nbrs>");
	    System.exit(1);
	}

	int m = Integer.parseInt(args[0]);
	int d = Integer.parseInt(args[1]);
	int n = Integer.parseInt(args[2]);

	double [][] keys = new double [m][d];

	double [] targ = new double [d];
	for (int k=0; k<d; ++k) {
	    targ[k] = 0.5;
	}

	// make a D-dimensional KD-tree
	KDTree<Integer> kd = new KDTree<Integer>(d);

	try {

	    // add M randomly keyed nodes

	    for (int i=0; i<m; ++i) {
		for (int j=0; j<d; ++j) {
		    keys[i][j] = r.nextDouble();
		}
		kd.insert(keys[i], i);
	    }

	    // get N nearest neighbors and show their keys

	    long start = time();

	    List<Integer> nbrs = kd.nearest(targ, n);
	
	    System.err.println((time() - start) + " msec");

	    //System.exit(0);

	    for (int j : nbrs) {
		for (int k=0; k<d; ++k) {
		    System.out.print(keys[j][k] + " ");
		}
		System.err.println();
	    }
	}
	catch (Exception e) {
	    System.err.println(e);
	}
    }	

    private static long time() {
	java.util.GregorianCalendar cal = new java.util.GregorianCalendar();
	return cal.getTimeInMillis();
    }



}


examples/kdexamples/kdtime.java
// illustrates differences between KD tree search and exhaustive search

import edu.wlu.cs.levy.CG.KDTree;

import java.util.Date;
import java.util.Random;

class kdtime {

    public static void main(String [] args) {

	// grab command-line args
	if (args.length < 3) {
	    System.err.print("Usage: java kdtime <# points> <# dims> ");
	    System.err.println("<# trials> [seed]");
	    System.exit(1);
	}
	int n = Integer.parseInt(args[0]);
	int k = Integer.parseInt(args[1]);
	int t = Integer.parseInt(args[2]);

	// generate N random K-dimensional points in (0,1)
	Random r = args.length > 3 ?               // support random seed as
	    new Random(Long.parseLong(args[3])) :  // optional fourth arg
	    new Random();
	double [][] x = new double[n][k];
	for (int i=0; i<n; ++i) {
	    for (int j=0; j<k; ++j) {
		x[i][j] = r.nextDouble();
	    }
	}

	// build KD-tree with indices as values
	KDTree<Integer> kd = new KDTree<Integer>(k);
	try {
	    for (int i=0; i<n; ++i) {
		kd.insert(x[i], i);
	    }
	}
	catch (Exception e) {
	    System.err.println(e);
	    System.exit(0);
	}

	// compare search
	linear_search(x, t, r);
	kd_search(x, t, kd, r);

	// compare nearest-neighbor
	linear_nearest(x, t, r);
	kd_nearest(x, t, kd, r);
    }

    // do linear search
    static void linear_search(double [][] x, int t, Random r) {
	
	long before = getTimeMillis();
	for (int i=0; i<t; ++i) {

	    // pick a random point
	    double [] targ = x[(int)(x.length*r.nextDouble())];
	    
	    search(x, targ);
	}
	long millis = getTimeMillis() - before;
	System.out.println(t + " linear searches took " + millis + " msec.");
    }

    
    // do KD-tree search
    static void kd_search(double [][] x, int t, KDTree kd, Random r) {

	long before = getTimeMillis();
	try {
	    for (int i=0; i<t; ++i) {
		
		// pick a random point
		double [] targ = x[(int)(x.length*r.nextDouble())];
		
		kd.search(targ);
	    }
	}
	catch (Exception e) {
	    System.err.println(e);
	    System.exit(0);
	}	
	long millis = getTimeMillis() - before;
	System.out.println(t + " KD-tree searches took " + millis + " msec.");
    }

	
    // do linear nearest neighbor
    static void linear_nearest(double [][] x, int t, Random r) {

	int k = x[0].length;
	long before = getTimeMillis();
	for (int i=0; i<t; ++i) {

	    double [] targ = random_point(r, k);
	    
	    int n = neighbor(x, targ);

	}
	long millis = getTimeMillis() - before;
	System.out.println(t + " linear nearest took " + millis + " msec.");
	
    }


    // do KD-tree nearest neighbor
    static void kd_nearest(double [][] x, int t, KDTree kd, Random r) {

	int k = x[0].length;
	long before = getTimeMillis();
	try {
	    for (int i=0; i<t; ++i) {
		
		double [] targ = random_point(r, k);
		Integer nbr = (Integer)kd.nearest(targ);
		int n = nbr.intValue();
	    }
	}
	catch (Exception e) {
	    System.err.println(e);
	    System.exit(0);
	}	
	long millis = getTimeMillis() - before;
	System.out.println(t + " KD-tree nearest took " + millis + " msec.");
    }
    // linear search for exact match
    private static int search(double [][] x, double [] targ) {

	for (int i=0; i<x.length; ++i) {
	    if (equal(x[i], targ)) return i;
	}

	return -1;
    }

    // linear search for index of neighbor
    private static int neighbor(double [][] x, double [] targ) {

	double mindst = Double.POSITIVE_INFINITY;
	int minidx = -1;

	for (int i=0; i<x.length; ++i) {
	    double d = sqrdst(x[i], targ);
	    if (d < mindst) {
		mindst = d;
		minidx = i;
	    }
	}
	
	return minidx;
    }

    // square of Euclidean distance between points
    private static double sqrdst(double [] p, double [] q) {
	double dst = 0;
	for (int i=0; i<p.length; ++i) {
	    double dif = p[i] - q[i];
	    dst += dif*dif;
	}
	return dst;
    }

    // point equality test
    private static boolean equal(double [] p, double [] q) {
	for (int i=0; i<p.length; ++i) {
	    if (p[i] != q[i]) return false;
	}
	return true;
    }

    private static double [] random_point(Random r, int k) {

	double [] x = new double[k];
	for (int i=0; i<k; ++i) {
	    x[i] = r.nextDouble();
	}
	return x;
    }

    private static long getTimeMillis() {
	Date d = new Date();
	return d.getTime();
    }
}


src/main/java/edu/wlu/cs/levy/CG/KDTree.java
package edu.wlu.cs.levy.CG;

import java.util.LinkedList;
import java.util.List;
import java.util.Stack;

/**
 * KDTree is a class supporting KD-tree insertion, deletion, equality search,
 * range search, and nearest neighbor(s) using double-precision floating-point
 * keys. Splitting dimension is chosen naively, by depth modulo K. Semantics are
 * as follows:
 *
* <UL>
 * <LI> Two different keys containing identical numbers should retrieve the same
 * value from a given KD-tree. Therefore keys are cloned when a node is
 * inserted.
 * <BR><BR>
 * <LI> As with Hashtables, values inserted into a KD-tree are <I>not</I>
 * cloned. Modifying a value between insertion and retrieval will therefore
 * modify the value stored in the tree.
 * </UL>
 *
 * Implements the Nearest Neighbor algorithm (Table 6.4) of
 *
* <PRE>
 * &#064;techreport{AndrewMooreNearestNeighbor,
 *   author  = {Andrew Moore},
 *   title   = {An introductory tutorial on kd-trees},
 *   institution = {Robotics Institute, Carnegie Mellon University},
 *   year    = {1991},
 *   number  = {Technical Report No. 209, Computer Laboratory,
 *              University of Cambridge},
 *   address = {Pittsburgh, PA}
 * }
 * </PRE>
 *
*
 * @author Simon Levy, Bjoern Heckel
 * @version %I%, %G%
 * @param <T>
 * @since JDK1.2
 */
public class KDTree<T> {

    // number of milliseconds
    final long m_timeout;

    // K = number of dimensions
    final private int m_K;

    // root of KD-tree
    private KDNode<T> m_root;

    // count of nodes
    private int m_count;

    /**
     * Creates a KD-tree with specified number of dimensions.
     *
     * @param k number of dimensions
     */
    public KDTree(int k) {
        this(k, 0);
    }

    public KDTree(int k, long timeout) {
        this.m_timeout = timeout;
        m_K = k;
        m_root = null;
    }

    /**
     * Insert a node in a KD-tree. Uses algorithm translated from 352.ins.c of
     *
     * <PRE>
     *   &#064;Book{GonnetBaezaYates1991,
     *     author =    {G.H. Gonnet and R. Baeza-Yates},
     *     title =     {Handbook of Algorithms and Data Structures},
     *     publisher = {Addison-Wesley},
     *     year =      {1991}
     *   }
     * </PRE>
     *
     * @param key key for KD-tree node
     * @param value value at that key
     *
     * @throws KeySizeException      if key.length mismatches K
     * @throws KeyDuplicateException if key already in tree
     */
    public void insert(double[] key, T value)
            throws KeySizeException, KeyDuplicateException {
        this.edit(key, new Editor.Inserter<T>(value));
    }

    /**
     * Edit a node in a KD-tree
     *
     * @param key key for KD-tree node
     * @param editor object to edit the value at that key
     *
     * @throws KeySizeException      if key.length mismatches K
     * @throws KeyDuplicateException if key already in tree
     */
    public void edit(double[] key, Editor<T> editor)
            throws KeySizeException, KeyDuplicateException {

        if (key.length != m_K) {
            throw new KeySizeException();
        }

        synchronized (this) {
            // the first insert has to be synchronized
            if (null == m_root) {
                m_root = KDNode.create(new HPoint(key), editor);
                m_count = m_root.deleted ? 0 : 1;
                return;
            }
        }

        m_count += KDNode.edit(new HPoint(key), editor, m_root, 0, m_K);
    }

    /**
     * Find KD-tree node whose key is identical to key. Uses algorithm
     * translated from 352.srch.c of Gonnet & Baeza-Yates.
     *
     * @param key key for KD-tree node
     *
     * @return object at key, or null if not found
     *
     * @throws KeySizeException if key.length mismatches K
     */
    public T search(double[] key) throws KeySizeException {

        if (key.length != m_K) {
            throw new KeySizeException();
        }

        KDNode<T> kd = KDNode.srch(new HPoint(key), m_root, m_K);

        return (kd == null ? null : kd.v);
    }

    public void delete(double[] key)
            throws KeySizeException, KeyMissingException {
        delete(key, false);
    }

    /**
     * Delete a node from a KD-tree. Instead of actually deleting node and
     * rebuilding tree, marks node as deleted. Hence, it is up to the caller to
     * rebuild the tree as needed for efficiency.
     *
     * @param key key for KD-tree node
     * @param optional if false and node not found, throw an exception
     *
     * @throws KeySizeException    if key.length mismatches K
     * @throws KeyMissingException if no node in tree has key
     */
    public void delete(double[] key, boolean optional)
            throws KeySizeException, KeyMissingException {

        if (key.length != m_K) {
            throw new KeySizeException();
        }
        KDNode<T> t = KDNode.srch(new HPoint(key), m_root, m_K);
        if (t == null) {
            if (optional == false) {
                throw new KeyMissingException();
            }
        } else {
            if (KDNode.del(t)) {
                m_count--;
            }
        }
    }

    /**
     * Find KD-tree node whose key is nearest neighbor to key.
     *
     * @param key key for KD-tree node
     *
     * @return object at node nearest to key, or null on failure
     *
     * @throws KeySizeException if key.length mismatches K
     *
     */
    public T nearest(double[] key) throws KeySizeException {

        List<T> nbrs = nearest(key, 1, null);
        return nbrs.get(0);
    }

    /**
     * Find KD-tree nodes whose keys are <i>n</i> nearest neighbors to key.
     *
     * @param key key for KD-tree node
     * @param n number of nodes to return
     *
     * @return objects at nodes nearest to key, or null on failure
     *
     * @throws KeySizeException if key.length mismatches K
     *
     */
    public List<T> nearest(double[] key, int n)
            throws KeySizeException, IllegalArgumentException {
        return nearest(key, n, null);
    }

    /**
     * Find KD-tree nodes whose keys are within a given Euclidean distance of a
     * given key.
     *
     * @param key key for KD-tree node
     * @param dist Euclidean distance
     *
     * @return objects at nodes with distance of key, or null on failure
     *
     * @throws KeySizeException if key.length mismatches K
     *
     */
    public List<T> nearestEuclidean(double[] key, double dist)
            throws KeySizeException {
        return nearestDistance(key, dist, new EuclideanDistance());
    }

    /**
     * Find KD-tree nodes whose keys are within a given Hamming distance of a
     * given key.
     *
     * @param key key for KD-tree node
     * @param dist Hamming distance
     *
     * @return objects at nodes with distance of key, or null on failure
     *
     * @throws KeySizeException if key.length mismatches K
     *
     */
    public List<T> nearestHamming(double[] key, double dist)
            throws KeySizeException {

        return nearestDistance(key, dist, new HammingDistance());
    }

    /**
     * Find KD-tree nodes whose keys are <I>n</I> nearest neighbors to key. Uses
     * algorithm above. Neighbors are returned in ascending order of distance to
     * key.
     *
     * @param key key for KD-tree node
     * @param n how many neighbors to find
     * @param checker an optional object to filter matches
     *
     * @return objects at node nearest to key, or null on failure
     *
     * @throws KeySizeException         if key.length mismatches K
     * @throws IllegalArgumentException if <I>n</I> is negative or exceeds tree
     *                                  size
     */
    public List<T> nearest(double[] key, int n, Checker<T> checker)
            throws KeySizeException, IllegalArgumentException {

        if (n <= 0) {
            return new LinkedList<T>();
        }

        NearestNeighborList<KDNode<T>> nnl = getnbrs(key, n, checker);

        n = nnl.getSize();
        Stack<T> nbrs = new Stack<T>();

        for (int i = 0; i < n; ++i) {
            KDNode<T> kd = nnl.removeHighest();
            nbrs.push(kd.v);
        }

        return nbrs;
    }

    /**
     * Range search in a KD-tree. Uses algorithm translated from 352.range.c of
     * Gonnet & Baeza-Yates.
     *
     * @param lowk lower-bounds for key
     * @param uppk upper-bounds for key
     *
     * @return array of Objects whose keys fall in range [lowk,uppk]
     *
     * @throws KeySizeException on mismatch among lowk.length, uppk.length, or K
     */
    public List<T> range(double[] lowk, double[] uppk)
            throws KeySizeException {

        if (lowk.length != uppk.length) {
            throw new KeySizeException();
        } else if (lowk.length != m_K) {
            throw new KeySizeException();
        } else {
            List<KDNode<T>> found = new LinkedList<KDNode<T>>();
            KDNode.rsearch(new HPoint(lowk), new HPoint(uppk),
                    m_root, 0, m_K, found);
            List<T> o = new LinkedList<T>();
            for (KDNode<T> node : found) {
                o.add(node.v);
            }
            return o;
        }
    }

    public int size() { /* added by MSL */

        return m_count;
    }

    public String toString() {
        return m_root.toString(0);
    }

    private NearestNeighborList<KDNode<T>> getnbrs(double[] key)
            throws KeySizeException {
        return getnbrs(key, m_count, null);
    }

    private NearestNeighborList<KDNode<T>> getnbrs(double[] key, int n,
            Checker<T> checker) throws KeySizeException {

        if (key.length != m_K) {
            throw new KeySizeException();
        }

        NearestNeighborList<KDNode<T>> nnl = new NearestNeighborList<KDNode<T>>(n);

        // initial call is with infinite hyper-rectangle and max distance
        HRect hr = HRect.infiniteHRect(key.length);
        double max_dist_sqd = Double.MAX_VALUE;
        HPoint keyp = new HPoint(key);

        if (m_count > 0) {
            long timeout = (this.m_timeout > 0)
                           ? (System.currentTimeMillis() + this.m_timeout)
                           : 0;
            KDNode.nnbr(m_root, keyp, hr, max_dist_sqd, 0, m_K, nnl, checker, timeout);
        }

        return nnl;

    }

    private List<T> nearestDistance(double[] key, double dist,
            DistanceMetric metric) throws KeySizeException {

        NearestNeighborList<KDNode<T>> nnl = getnbrs(key);
        int n = nnl.getSize();
        Stack<T> nbrs = new Stack<T>();

        for (int i = 0; i < n; ++i) {
            KDNode<T> kd = nnl.removeHighest();
            HPoint p = kd.k;
            if (metric.distance(kd.k.coord, key) < dist) {
                nbrs.push(kd.v);
            }
        }

        return nbrs;
    }

}


edu/wlu/cs/levy/CG/KDTests.java
package edu.wlu.cs.levy.CG;


/*
  written by MSL for SpeedDate
*/
import java.util.List;


import org.junit.Test;
import org.junit.Ignore;
import org.junit.Assert;

public class KDTests {
    static java.util.Random rand = new java.util.Random();

    static double[] makeSample(int dims) {
        double [] rv = new double [dims];
        for (int j=0; j < dims; ++j) {
            rv[j] = rand.nextDouble();
        }
        return rv;
    }

    static double distSquared(double [] p0, double [] p1) {
        double rv = 0;
        for (int i=0; i < p0.length; i++) {
            double diff = p0[i] - p1[i];
            rv += (diff * diff);
        }
        return rv;
    }


    @Test public void testNearestNeighborList() {
        NearestNeighborList<String> nnl = new NearestNeighborList<String>(3);
        nnl.insert("A", 3.0);
        nnl.insert("B", 2.0);
        nnl.insert("D", 0.0);
        nnl.insert("C", 1.0);

        Assert.assertEquals(2.0, nnl.getMaxPriority(), 0.1);
        Assert.assertEquals("B", nnl.getHighest());
        Assert.assertEquals("B", nnl.removeHighest());
        Assert.assertEquals("C", nnl.removeHighest());
        Assert.assertEquals("D", nnl.removeHighest());
    }

    @Test public void testNearestNeighbor() throws KDException {
        int dims = 3;
        int samples = 300;
        KDTree<Integer> kt = new KDTree<Integer>(dims);
        double [] targ = makeSample(dims);

        int min_index = 0;
        double min_value = Double.MAX_VALUE;
        for (int i=0; i<samples; ++i) {
            double [] keys = makeSample(dims);
            kt.insert(keys, new Integer(i));

            /*
              for the purposes of test, we want the nearest EVEN-NUMBERED point 
            */
            if ( (i % 2) == 0) {
                double dist = distSquared(targ, keys);
                if (dist < min_value) {
                    min_value = dist;
                    min_index = i;
                }
            }
        }


        List<Integer> nbrs = kt.nearest(targ, 1, new Checker<Integer>() {
                public boolean usable(Integer v) {
                    return (v.intValue() % 2) == 0;
                }});
	
        Assert.assertEquals(1, nbrs.size());
        if (nbrs.size() == 1) {
            Assert.assertEquals(min_index, nbrs.get(0).intValue());
        }
    }

    @Test public void testRange() throws KDException {
        int dims = 2;
        KDTree<Object> kt = new KDTree<Object>(dims);
        double [] p0 = { 0.5, 0.5 };
        double [] p1 = { 0.65, 0.5 };
        double [] p2 = { 0.75, 0.5 };

        kt.insert(p0, new Object());
        kt.insert(p1, new Object());
        kt.insert(p2, new Object());

        double [] lower = { 0.25, 0.3 };
        double [] upper = { 0.7, 0.6 };

        List<Object> rv = kt.range(lower, upper);
        Assert.assertEquals(2, rv.size());

        kt.delete(p1);
        rv = kt.range(lower, upper);
        Assert.assertEquals(1, rv.size());
    }

    @Test public void testSearch() throws KDException {
        int dims = 3;
        int samples = 300;
        KDTree<Object> kt = new KDTree<Object>(dims);
        double [] targ = makeSample(dims);
        Object treasure = new Object();
        kt.insert(targ, treasure);

        for (int i=0; i<samples; ++i) {
            double [] keys = makeSample(dims);
            kt.insert(keys, new Integer(i));
        }

        Object found = kt.search(targ);
        Assert.assertSame(treasure, found);

        kt.delete(targ);
        found = kt.search(targ);
        Assert.assertNull(found);

    }

    @Test public void testDelete() throws KDException {
        int dims = 3;
        KDTree<Object> kt = new KDTree<Object>(dims);
        double [] targ = makeSample(dims);
        kt.insert(targ, new Object());
        kt.delete(targ);
        try {
            kt.delete(targ);
            Assert.assertFalse(true);
        } catch (edu.wlu.cs.levy.CG.KeyMissingException e) { 
            // supposed to be here
        }
        kt.delete(targ, true);
        Assert.assertEquals(0, kt.size());
    }

    @Test public void testEditing() throws KDException {
        int dims = 3;
        KDTree<Object> kt = new KDTree<Object>(dims);
        double [] targ = makeSample(dims);

        Object p1 = "p1";
        Object p2 = "p2";
        kt.insert(targ, p1);
        try {
            kt.insert(targ, p2);
            Assert.assertFalse(true);
        } catch (edu.wlu.cs.levy.CG.KeyDuplicateException e) { 
            // supposed to be here
        }

        kt.edit(targ, new Editor.OptionalInserter<Object>(p2));
        Object found = kt.search(targ);
        Assert.assertSame(p1, found);
        kt.edit(targ, new Editor.Replacer<Object>(p2));
        found = kt.search(targ);
        Assert.assertSame(p2, found);
    }
}
